cmd/snap-bootstrap/cmd_initramfs_mounts.go

```diff
@@ -269,29 +285,176 @@ func copyFromGlobHelper(src, dst, globEx string) error {
 	return nil
 }
 
-func generateMountsModeRecover(mst *initramfsMountsState) error {
-	// steps 1 and 2 are shared with install mode
-	if err := generateMountsCommonInstallRecover(mst); err != nil {
-		return err
-	}
+type recoverDegradedState struct {
+	// TODO: make these values enums?
+	// DataKey is which key was used to unlock ubuntu-data (if any). Valid
+	// values are:
+	// - "run" for the normal run mode key object
+	// - "fallback" for the fallback recover mode specific key object
+	// - "" for either unencrypted case, or if we were unable to unlock it
+	//   (in the case we failed to unlock it, but we know it is there,
+	//   the DataState field provides more information)
+	DataKey string `json:"data-key"`
+	// TODO: make these values enums?
+	// DataState is the state of the ubuntu-data mountpoint, it can be:
+	// - "mounted" for mounted and available
+	// - "enc-not-found" for state where we found an ubuntu-data-enc
+	//   partition and tried to unlock it, but failed _before mounting it_.
+	//   Errors mounting are identified differently.
+	// TODO:UC20: "other" is a really bad term here replace with something better
+	// - "other" for some other error where we couldn't identify if there's an
+	//    ubuntu-data (or ubuntu-data-enc) partition at all
+	DataState string `json:"data-state"`
+	// SaveKey is which key was used to unlock ubuntu-save (if any). Same values
+	// as DataKey.
+	SaveKey string `json:"save-key"`
+	// SaveState is the state of the ubuntu-save mountpoint, it can be in any of
+	// the states documented for DataState, with the additional state of:
+	// - "not-found" for state when we don't have ubuntu-save, but it is
+	//   optional (unencrypted only)
+	SaveState string `json:"save-state"`
+	// HostLocation is the location where the host's ubuntu-data is mounted
+	// and available. If this is the empty string, then the host's
+	// ubuntu-data is not available anywhere.
+	HostLocation string `json:"host-location"`
+	// ErrorLog holds the log of errors that led to the degraded recovery mode.
+	ErrorLog []string `json:"error-log"`
+}
 
-	// get the disk that we mounted the ubuntu-seed partition from as a
-	// reference point for future mounts
-	disk, err := disks.DiskFromMountPoint(boot.InitramfsUbuntuSeedDir, nil)
+func (r *recoverDegradedState) LogErrorf(format string, v ...interface{}) {
+	msg := fmt.Sprintf(format, v...)
+	r.ErrorLog = append(r.ErrorLog, msg)
+	logger.Noticef(msg)
+}
+
+type recoverContext struct {
+	// the disk we have all our partitions on
+	disk disks.Disk
+	// options for finding partitions on the disk we have all our partitions on
+	diskOpts *disks.Options
+	// the device for ubuntu-data that is to be mounted - either an unencrypted
+	// volume, or a decrypted mapper volume if ubuntu-data is really encrypted
+	// on the physical disk
+	dataDevice string
+	// the device for ubuntu-save that is to be mounted
+	saveDevice string
+
+	// XXX: would be nice to get rid of tracking this here
```

Comment:
it just feels awkward to track this in the ctx instead of the states themselves knowing about whether the disk is an encrypted one or not, i.e. we have some states like `locateUnencryptedSave` that are clearly only for unencrypted disks, and some states that are clearly only for encrypted like `unlockSaveFallbackKey` it just feels like instead of tracking more state in the state machine that we could have better states, but I think this is fine as-is

Reply:
this is probably okay now